packages/go/go-helpers.ts

```diff
@@ -112,72 +112,146 @@ class GoWrapper {
   }
 }
 
-export async function createGo(
-  workPath: string,
-  goPath: string,
-  platform = process.platform,
-  arch = process.arch,
-  opts: execa.Options = {},
-  goMod = false
-) {
-  const binPath = join(getGoDir(workPath), 'bin');
-  debug(`Adding ${binPath} to PATH`);
-  const path = `${binPath}${delimiter}${process.env.PATH}`;
-  const env: { [key: string]: string } = {
-    ...process.env,
-    PATH: path,
-    GOPATH: goPath,
-    ...opts.env,
-  };
-  if (goMod) {
-    env.GO111MODULE = 'on';
+type CreateGoOptions = {
+  goPath?: string;
+  modulePath?: string;
+  opts?: execa.Options;
+  workPath: string;
+};
+
+export async function createGo({
+  goPath,
+  modulePath,
+  opts = {},
+  workPath,
+}: CreateGoOptions) {
+  if (goPath === undefined) {
+    goPath = getGoDir(workPath);
   }
-  await createGoPathTree(goPath, platform, arch);
-  return new GoWrapper(env, opts);
-}
 
-export async function downloadGo(workPath: string, modulePath: string) {
-  const dir = getGoDir(workPath);
-  const { platform, arch } = process;
-  const version = await parseGoVersion(modulePath);
+  // parse the `go.mod`, if exists
+  let goPreferredVersion;
+  if (modulePath) {
+    goPreferredVersion = await parseGoModVersion(modulePath);
+  }
 
-  // Check if `go` is already installed in user's `$PATH`
-  const { failed, stdout } = await execa('go', ['version'], { reject: false });
+  // default to newest (first) supported go version
+  const goSelectedVersion =
+    goPreferredVersion || Array.from(versionMap.values())[0];
 
-  if (!failed && parseInt(stdout.split('.')[1]) >= GO_MIN_VERSION) {
-    debug('Using system installed version of `go`: %o', stdout.trim());
-    return createGo(workPath, dir, platform, arch);
-  }
+  const env = cloneEnv(process.env, opts.env);
+  const { arch, platform } = process;
+  const goGlobalDir = join(
+    goGlobalCachePath,
+    `${goSelectedVersion}_${platform}_${arch}`
+  );
+  const goGlobalBinDir = join(goGlobalDir, 'bin');
 
-  // Check `go` bin in cacheDir
-  const isGoExist = await pathExists(join(dir, 'bin'));
-  if (!isGoExist) {
-    debug('Installing `go` v%s to %o for %s %s', version, dir, platform, arch);
-    const url = getGoUrl(version, platform, arch);
-    debug('Downloading `go` URL: %o', url);
-    const res = await fetch(url);
+  if (goPreferredVersion) {
+    debug(`Preferred go version ${goPreferredVersion} (from go.mod)`);
+    env.GO111MODULE = 'on';
+  }
 
-    if (!res.ok) {
-      throw new Error(`Failed to download: ${url} (${res.status})`);
+  // check we have the desired `go` version cached
+  if (await pathExists(goGlobalBinDir)) {
+    // check if `go` has already been downloaded and that the version is correct
+    env.GOROOT = goGlobalDir;
+    env.PATH = `${goGlobalBinDir}${delimiter}${env.PATH}`;
+    const { failed, stdout } = await execa('go', ['version'], {
+      env,
+      reject: false,
+    });
+    if (!failed) {
```

Comment:
**suggestions (non-blocking):** Can we debug log when this does fail?

Reply:
Originally, I had logging in there, but it was making the logs too verbose. In this case, it doesn't really matter that it failed because it'll still try the system PATH branch before downloading the correct version.